pkg/controllers/deprovisioning/consolidation.go

```diff
@@ -58,106 +71,146 @@ func (c *Consolidation) ShouldDeprovision(_ context.Context, n *state.Node, prov
 	return provisioner != nil && provisioner.Spec.Consolidation != nil && ptr.BoolValue(provisioner.Spec.Consolidation.Enabled)
 }
 
-// sortCandidates orders deprovisionable nodes by the disruptionCost
-func (c *Consolidation) SortCandidates(nodes []CandidateNode) []CandidateNode {
-	sort.Slice(nodes, func(i int, j int) bool {
-		return nodes[i].disruptionCost < nodes[j].disruptionCost
-	})
-	return nodes
-}
-
 // computeCommand generates a deprovisioning command given deprovisionable nodes
-func (c *Consolidation) ComputeCommand(ctx context.Context, attempt int, candidates ...CandidateNode) (Command, error) {
+func (c *Consolidation) ComputeCommand(ctx context.Context, candidates ...CandidateNode) (Command, error) {
 	// the last cluster consolidation wasn't able to improve things and nothing has changed regarding
 	// the cluster that makes us think we would be successful now
 	if c.lastConsolidationState == c.cluster.ClusterConsolidationState() {
 		return Command{action: actionDoNothing}, nil
 	}
+	// First sort the candidate nodes by disruption cost.
+	sort.Slice(candidates, func(i int, j int) bool {
+		return candidates[i].disruptionCost < candidates[j].disruptionCost
+	})
+
+	// We use this timestamp for the Consolidation TTL so that our TTL logic waits one time per consolidation loop.
+	validationTimestamp := time.Time{}
+	var cmd Command
+	var err error
+
 	// First delete any empty nodes we see
-	if attempt == 0 {
-		if cmd := c.deleteEmpty(ctx, candidates...); cmd.action == actionDelete {
-			return cmd, nil
-		}
+	cmd, validationTimestamp, err = c.deleteEmpty(ctx, validationTimestamp, candidates...)
+	if err != nil {
+		return Command{action: actionFailed}, fmt.Errorf("deleting empty nodes for consolidation, %w", err)
+	}
+	// If there are empty nodes, delete the empty nodes for this provisioning loop.
+	if cmd.action == actionDelete {
+		return cmd, nil
 	}
 
 	pdbs, err := NewPDBLimits(ctx, c.kubeClient)
 	if err != nil {
 		return Command{action: actionFailed}, fmt.Errorf("tracking PodDisruptionBudgets, %w", err)
 	}
-	// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
-	// scheduling unless its really needed
-	if !canBeTerminated(candidates[attempt], pdbs) {
-		return Command{action: actionNotPossible}, nil
-	}
 
-	return c.computeConsolidation(ctx, candidates[attempt])
+	for _, candidate := range candidates {
+		// is this a node that we can terminate?  This check is meant to be fast so we can save the expense of simulated
+		// scheduling unless its really needed
+		if !canBeTerminated(candidate, pdbs) {
+			continue
+		}
+		// do a simulated scheduling loop to check if Consolidation is possible
+		cmd, validationTimestamp, err = c.computeConsolidation(ctx, validationTimestamp, candidate)
```

Comment:
Can we get rid of this `validationTimestamp`, it's threaded through a bunch of stuff but I don't think it's necessary.  You could just capture the start time before you iterate over the candidates.  When you wait for the TTL, the time you've already waited is the time since then.  This lets you skip the wait on subsequent commands and should remove a decent amount of code.

Reply:
I guess we could, just means if we're stuck computing for too long, it doesn't necessarily always work, but happy to pull out.

Reply:
Done